### PR TITLE
Fix issues: color codes and loc with scalars

### PR DESCRIPTION
**Pull request by eebasso**
Dark+ Theme
Screenshot 2023-06-24 at 3 42 00 AM

Night Owl Theme
Screenshot 2023-06-24 at 3 42 14 AM

Stellaris Theme
Screenshot 2023-06-24 at 3 42 26 AM

BEGIN_COMMIT_OVERRIDE
feat: added new color codes

fix: fixed color codes without a matching close tag affecting the rest of the words
fix: fixed color highlighting not working with scalars 
END_COMMIT_OVERRIDE

### CODE REVIEW

File: themes/dark.json [hunk 1]
```diff
@@ -570,11 +569,11 @@
 		},
 		{
 			"scope": [
-				"keyword.bracket-command.stellaris.yml",
-				"keyword.loc-key.stellaris.yml"
+				"bracket-command.stellaris.yml",
+				"loc-key.stellaris.yml"
```
Comment (The24thDS) [inline themes/dark.json:569]: What is the reason behind this change?

File: themes/stellaris_theme.json [hunk 2]
```diff
@@ -239,7 +300,7 @@
       }
     },
     {
-      "scope": "entity.name.tag.yaml",
+      "scope": "entity.name.tag",
```
Comment (The24thDS) [inline themes/stellaris_theme.json]: Wouldn't this change make it so this color is applied to all tags, in all languages, not just yaml?

**Review by The24thDS — COMMENTED** [review]
Hello 👋 

First of all, let me thank you for this PR! I didn't expect contributions but they are very welcome :) 
I'm going to take a better look at this over the next week but do you mind leaving a comment outlining the changes that you made and explaining how you did them? It will make the review process faster. 
I left a couple comments in some areas that peeked my interest right now but I might leave more over the week.

Once again, thank you for this! Let's work together to get this shipped to the users as soon as possible :)

File: themes/dark.json (hunk 1 above)
Comment (eebasso) [reply themes/dark.json:569]: With keyword.loc-key.stellaris, you'll notice that things like `$some_loc_key$` were always the same color as "keyword," even when when it was nested between §H and §!, for example. This is because it was picking up the color from the "keyword" scope instead of the color of the scope "keyword.color-code-h.stellaris." I found that this change fixed that.

File: themes/stellaris_theme.json (hunk 2 above)
Comment (eebasso) [reply themes/stellaris_theme.json]: Not sure, but I wasn't expecting this theme to be used with any other language. I use the extension "Theme by language" so I've only looked at how this theme looks in YAML. A more specific alternative could be 
"scope": [ "entity.name.tag.yaml", "entity.name.tag.stellaris.yaml" ]

**Comment by eebasso** [comment]
Thank you for making this very useful syntax and themes, and you are quite welcome for the contributions :). When I tried to leave comments in the code, VSCode yelled at me saying no comments in JSON files.

For the colorcodes, I basically added a bunch of lookaheads to the “begin” of the syntax rules so that it would require the appropriate ending to occur on the same line.

I changed the brackets and loc-key so that not only no spaces occur, but also no ‘$’, ‘[‘, ‘]’ or new lines.

I tweaked the scopes and theme so that loc-key and bracket would adopt the same color as keyword.colorcode instead of keyword, and also appear both bold and underlined. The trade off was that they will appear white when outside of a colorcode scope, but I think that’s arguably better since it indicates the true color in game and the bold underline is enough imo to distinguish it.

I added new colorcodes from the latest updates and DLCs.

File: themes/dark.json (hunk 1 above)
Comment (The24thDS) [reply themes/dark.json:569]: I like this change, we can see the color but we still get the underline

File: themes/stellaris_theme.json (hunk 2 above)
Comment (The24thDS) [reply themes/stellaris_theme.json]: I say let's keep the `entity.name.tag.yaml` scope that we had before. We can't assume that everyone is using that extension and we should try to minimize our dependency on other extensions.

**Comment by The24thDS** [comment]
I'm seeing a bug where a color code doesn't seem to apply anymore in this situation. I tried removing some of the other keywords and reducing the content to see if I can narrow done the problem but with no success.
It will work if you add a closing "tag" (§!) but this is not indicative of how it will behave in game.
image

**Comment by The24thDS** [comment]
You gave me the motivation required to find a way of making this match properly.
I came up with the following:
```json
    "color-code-w": {
      "begin": "§W",
      "end": "§!|(?=\"$)",
      "name": "keyword.color-code.w.stellaris.yml"
    }
 ```
This seems to be working as we want it to. It stops at the end of the string if it can't find a matching close tag.
image
Please put it in your branch and test a bit as well :)

File: package.json [hunk 3]
```diff
@@ -57,5 +64,15 @@
...
+  "__metadata": {
+    "id": "3c607bba-2b85-49c0-afbf-f1e99e92b257",
+    "publisherId": "f63e6cf8-df88-48c9-9d40-9ab9bda726e7",
+    "publisherDisplayName": "The24thDS",
+    "targetPlatform": "undefined",
+    "updated": false,
+    "isPreReleaseVersion": false,
+    "installedTimestamp": 1687544745173,
+    "preRelease": false
+  }
```
Comment (The24thDS) [inline package.json:77]: How did these get added? 😅

File: themes/dark.json [hunk 4]
```diff
@@ -450,7 +450,6 @@
 		},
 		{
 			"scope": [
-				"keyword.control",
```
Comment (The24thDS) [inline themes/dark.json:453]: Any reason for removing this one?

**Review by The24thDS — CHANGES_REQUESTED** [review]
I left some more comments and suggested a change to how we match color codes above.

Some things to do after we got the matching working properly:
- update the readme with new images of the themes;
- remove the fixed items of the `Known issues` list;
- update the `Extensions settings` section with the new codes.

File: package.json (hunk 3 above)
Comment (eebasso) [reply package.json:77]: I found that in the files coming from the VSCode extension market. I added it on a whim really, I have no idea if it's necessary or useful haha.

**Comment by eebasso** [comment]
> I'm seeing a bug where a color code doesn't seem to apply anymore in this situation. I tried removing some of the other keywords and reducing the content to see if I can narrow done the problem but with no success. It will work if you add a closing "tag" (§!) but this is not indicative of how it will behave in game. image

Yeah it's interesting how it stills works in game even with the "incorrect" syntax. I think your idea of making an alternative ending of a quotation is a good idea. In fact, I had come up with something similar independently. Great minds :-)

I'll upload the update to the branch soon.

Edit: One point I will make in favor of keeping it like this is the following: **it helps users catch the syntax error of not closing with "§!" as the user will notice that the color does not appear.** If you look at how Paradox does it in their localisation files, they seem to be 100% consistent on that syntax rule of closing with §!. Therefore, it could perhaps be better for the user if the color doesn't appear even though it technically would appear in the game. Something to consider...

File: themes/dark.json (hunk 4 above)
Comment (eebasso) [reply themes/dark.json:453]: That's because the scope "keyword.control" already shows up in that theme on line 316, and I liked the color there (light blue "#569CD6") vs the one removed on 453 (pinkish "#C586C0"). We can switch it if you like the other color better, but I think we shouldn't have both to avoid bugs and ambiguities. Generally I think each scope should only map to one color.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,9 +2,22 @@
...
+### Fixed issues
+
+- color codes without a matching close tag will no longer affect new lines of the document (add commit/PR here)
+- locs now work with scalars (add commit/PR here)
+
+### Color code updates
+
+- references are now same scoped color as neighboring text (add commit/PR here)
+- added new color codes from latest Stellaris updates (add commit/PR here)
+
```
Comment (The24thDS) [inline CHANGELOG.md:16]: Changelog gets generated automatically when we upload a new version

File: package.json (hunk 3 above)
Comment (The24thDS) [reply package.json:77]: We should probably remove it then. It looks like client side info since it has the installed time there

File: themes/dark.json (hunk 4 above)
Comment (The24thDS) [reply themes/dark.json:453]: Most of what is in these files come from the themes themselves. I'm surprised you went through them :D

File: syntaxes/stellaris_yaml_injection.json [hunk 6]
```diff
@@ -39,120 +39,120 @@
...
     },
     "icon-code": {
-      "match": "\\£[^\\s\\£\\n]+\\£",
+      "match": "\\£[^\\s\\n\\£]+\\£",
       "name": "keyword.icon-code.stellaris.yml"
     },
     "color-code-w": {
-      "begin": "§W(?=[^\\n]*§!)",
-      "end": "§!",
+      "begin": "§W(?=[^\\n]*(§!|(?=\"\\n)))",
```
Comment (The24thDS) [inline syntaxes/stellaris_yaml_injection.json]: I don't think we need this complex expression anymore. The way it is now it basically a `match` but it is better to use a simple `begin` and `end` and let VSCode figure out the magic :D If we use a `match` we would have to find a way to account for nested color codes

**Comment by The24thDS** [comment]
@eebasso are you still interested in making these changes?

**Comment by eebasso** [comment]
Oh for sure I'm interested. Sorry for being away, I've been busy moving.

File: syntaxes/stellaris_yaml_injection.json (hunk 6 above)
Comment (eebasso) [reply syntaxes/stellaris_yaml_injection.json]: From my experience, the complex expressions are necessary to avoid the color spilling over into other lines. VSCode gets it wrong a lot of the time, especially when there are nested color codes. I am running some tests now to see if the "begin" and/or "end" can be simplified.

Edit: seems like we can simplify "begin" but not "end." And you are right, using a match would make things harder to account for nested color codes.

Edit 2: I've updated the branch to the simple "begin" expressions.

Comment (The24thDS) [reply syntaxes/stellaris_yaml_injection.json]: Seems to be working as expected now

**Comment by The24thDS** [comment]
Is it possible that the new changes you made broke the scalar parsing? I'm no longer getting the colors if I add a scalar.
image

**EDIT:** 

Oh, I see what the problem is in that case: the line is indented by two spaces instead of one.
So not an actual problem since loc keys are supposed to be indented with one space only.

**Comment by The24thDS** [comment]
Oh, I see what the problem is in that case: the line is indented by two spaces instead of one.

**Comment by The24thDS** [comment]
I think we are almost ready to merge this. Just a few things that need to be done.

- remove the stuff from the `package.json` which I mentioned here https://github.com/The24thDS/stellaris-yaml/pull/1#discussion_r1245599843;
- update the  `Extension Settings` section of the `README.md` file with the new tokens;
- delete the release notes you wrote in `README.md` and `CHANGELOG.md`.

_Regarding release notes:_
- it is a good practice to follow the [conventional-commits](https://www.conventionalcommits.org/en/v1.0.0) rules, this way we can automatically generate release notes;
- because you have not followed the convention with your commits I will have to squash the PR into a single commit (unfortunately this means that we won't be able to automatically show nice release notes that would mention all the changes);
- while generating the new version I will manually change the release notes to better illustrate the changes brought by this PR.

**Comment by The24thDS** [comment]
Hi! I decided to merge this as in this and fix those small issues myself. Thank you for your contribution!
